Merge tag 'swift-DEVELOPMENT-SNAPSHOT-2019-09-24-a' into tensorflow

dan-zheng opened the pull request:
Recreation of https://github.com/apple/swift/pull/27397, merging from `apple/tensorflow-merge` branch.

gottesmm committed:
[builtin] Implement polymorphic builtins for all BUILTIN_BINARY_OPERATIONs.

TLDR: This patch introduces a new kind of builtin, "a polymorphic builtin". One
calls it like any other builtin, e.x.:

```
Builtin.generic_add(x, y)
```

but it has a contract: it must be specialized to a concrete builtin by the time
we hit Lowered SIL. In this commit, I add support for the following generic
operations:

Type           | Op
------------------------
FloatOrVector  |FAdd
FloatOrVector  |FDiv
FloatOrVector  |FMul
FloatOrVector  |FRem
FloatOrVector  |FSub
IntegerOrVector|AShr
IntegerOrVector|Add
IntegerOrVector|And
IntegerOrVector|ExactSDiv
IntegerOrVector|ExactUDiv
IntegerOrVector|LShr
IntegerOrVector|Mul
IntegerOrVector|Or
IntegerOrVector|SDiv
IntegerOrVector|SRem
IntegerOrVector|Shl
IntegerOrVector|Sub
IntegerOrVector|UDiv
IntegerOrVector|Xor
Integer        |URem

NOTE: I only implemented support for the builtins in SIL and in SILGen. I am
going to implement the optimizer parts of this in a separate series of commits.

DISCUSSION
----------

Today there are polymorphic like instructions in LLVM-IR. Yet, at the
swift and SIL level we represent these operations instead as Builtins whose
names are resolved by splatting the builtin into the name. For example, adding
two things in LLVM:

```
  %2 = add i64 %0, %1
  %2 = add <2 x i64> %0, %1
  %2 = add <4 x i64> %0, %1
  %2 = add <8 x i64> %0, %1
```

Each of the add operations are done by the same polymorphic instruction. In
constrast, we splat out these Builtins in swift today, i.e.:

```
let x, y: Builtin.Int32
Builtin.add_Int32(x, y)
let x, y: Builtin.Vec4xInt32
Builtin.add_Vec4xInt32(x, y)
...
```

In SIL, we translate these verbatim and then IRGen just lowers them to the
appropriate polymorphic instruction. Beyond being verbose, these prevent these
Builtins (which need static types) from being used in polymorphic contexts where
we can guarantee that eventually a static type will be provided.

In contrast, the polymorphic builtins introduced in this commit can be passed
any type, with the proviso that the expert user using this feature can guarantee
that before we reach Lowered SIL, the generic_add has been eliminated. This is
enforced by IRGen asserting if passed such a builtin and by the SILVerifier
checking that the underlying builtin is never called once the module is in
Lowered SIL.

In forthcoming commits, I am going to add two optimizations that give the stdlib
tool writer the tools needed to use this builtin:

1. I am going to add an optimization to constant propagation that changes a
"generic_*" op to the type of its argument if the argument is a type that is
valid for the builtin (i.e. integer or vector).

2. I am going to teach the SILCloner how to specialize these as it inlines. This
ensures that when we transparent inline, we specialize the builtin automatically
and can then form SSA at -Onone using predictable memory access operations.

The main implication around these polymorphic builtins are that if an author is
not able to specialize the builtin, they need to ensure that after constant
propagation, the generic builtin has been DCEed. The general rules are that the
-Onone optimizer will constant fold branches with constant integer operands. So
if one can use a bool of some sort to trigger the operation, one can be
guaranteed that the code will not codegen. I am considering putting in some sort
of diagnostic to ensure that the stdlib writer has a good experience (e.x. get
an error instead of crashing the compiler).

owenv committed:
Improve diagnostic when attempting to pass an Array to a variadic argument

- Give a more specific diagnostic which indicates the parameter is variadic
- If the argument is an Array literal, offer to drop the brackets

xedin committed:
[Diagnostics] Tailored diagnostic for "condition" expression

Since all condition expressions supposed to be convertible
to `Bool`, let's use that type as contextual and produce a
tailored diagnostic.

slavapestov committed:
Sema: Add a FIXME comment for a problem I found by inspection

I believe this is related to <https://bugs.swift.org/browse/SR-6218>
and <https://bugs.swift.org/browse/SR-8247>.

CodaFi committed:
Define UnderlyingTypeRequest

Define a request for computing the interface type of the underlying type of a typealias.  This can be used in place of the declared interface type of the alias itself when it is nested to preserve the fragile validation ordering issues that can cause.

CodaFi committed:
Resugar the structural type

Structural type computation ought to actually return a typealias type here using the same mechanism we compute the interface type with (minus the metatype).

CodaFi committed:
Improve the TypeAlias{Type|Decl} dump

Dump the underlying type by desugaring ourselves a little.

CodaFi committed:
Separate computing interface types and underlying types

Computing the interface type of a typealias used to push validation forward and recompute the interface type on the fly.  This was fragile and inconsistent with the way interface types are computed in the rest of the decls.  Separate these two notions, and plumb through explicit interface type computations with the same "computeType" idiom.  This will better allow us to identify the places where we have to force an interface type computation.

Also remove access to the underlying type loc.  It's now just a cache location the underlying type request will use.  Push a type repr accessor to the places that need it, and push the underlying type accessor for everywhere else.  Getting the structural type is still preferred for pre-validated computations.

This required the resetting of a number of places where we were - in many cases tacitly - asking the question "does the interface type exist".  This enables the removal of validateDeclForNameLookup

CodaFi committed:
Remove validateDeclForNameLookup

Flush it and the early validation hack now that we can delay computing the underlying interface type on demand and have taught type resolution to honor the structural type of a typealias.

This changes the way requirement signatures are spelled as a side effect.

CodaFi committed:
Setting the underlying type should set the structural type too

This avoids making the structural type dependent on whether the interface type has been computed and neatly avoids special-casing non-parsed declarations which set their underlying type up front.

gottesmm committed:
[gardening] Clean up BranchPropagatedUser.

ravikandhadai committed:
[OSLog][Test] Update the new oslog overlay implementation to use
@_semantics("constant_evaluable") annotation to denote constant
evaluable functions.

Add a test suite that uses the sil-opt pass ConstantEvaluableSubsetChecker.cpp
to check the constant evaluability of function in the OSLog
overlay.

nkcsgexi committed:
ABI/API checker: move node mapping information from under -v to under -debug. NFC

ravikandhadai committed:
[Constant Evaluator] Add a new sil-opt pass to check constant evaluability
of Swift code snippets. Add a new test: constant_evaluable_subset_test.swift
that tests Swift code snippets that are expected to be constant evaluable and
those that are not.

compnerd committed:
Revert "runtime: add a workaround for Windows build"

This reverts commit efaf1fbefa7fc94ff535b35b964689735637458b.
Add a much more palatable workaround for the unit tests.  Rather than
adding the dllimport for the symbols, locally define the required
symbols.  This list is sufficient to restore the ability to build tests
for Windows.

Catfish-Man committed:
Merge pull request #27203 from CodaFi/movin-on-up

Revert #26937, #27190, and #27132

Commit:
Ensure that Decls aren't added needlessly before creating tree eagerly.

xedin committed:
[Diagnostics] Don't store contextual type in missing argument fix/diagnostic

Instead of storing contextual function type in the fix/diagnostic,
let's fetch it from context (solution and/or locator) because it's
only used when it is a trailing closure missing some arguments anyway.

theblixguy committed:
[Test] Adds some tests

theblixguy committed:
[AST] Fix indenation in hasValueSemantics()

CodaFi committed:
Merge pull request #27080 from CodaFi/can-i-have-your-autograph

Requestify generic signatures

nkcsgexi committed:
Merge pull request #27200 from nkcsgexi/debug-flag

ABI/API checker: move node mapping information from under -v to under -debug. NFC

jckarter committed:
Merge pull request #27202 from jckarter/opaque-type-fallback-fix

IRGen: Use correct generic sig when mangling opaque underlying type.

beccadax committed:
[NFC] Add and use ASTContext::getNSCopyingDecl()

gottesmm committed:
[ownership] Create a context structure for the linear lifetime checker.

This is just a simple refactoring commit in preparation for hiding more of the
details of the linear lifetime checker. This is NFC, just moving around code.

swift-ci committed:
Merge pull request #27197 from gottesmm/pr-32d08675e348d877f8935092770eccd63eb67427

xedin committed:
[Diagnostics] Cover single explicit tuple argument -> N parameters via missing arguments fix

Expand tuple into N arguments to cover expect parameters.

This covers cases like this:

```swift
func foo(_: (Int, Int) -> Void) {}
foo { (bar: (Int, Int)) in } // used a single tuple type `(Int, Int)`
                             // instead of two distinct arguments
```

theblixguy committed:
Merge pull request #27198 from theblixguy/fix/SR-11478

[PropertyWrapper] Fix a bug with static property wrapper being rejected in a class

hamishknight committed:
[CS] Allow simplification of key path subscript arg locators (#27194)

[CS] Allow simplification of key path subscript arg locators

compnerd committed:
Merge pull request #27204 from compnerd/workaround

Revert "runtime: add a workaround for Windows build"

gottesmm committed:
Merge pull request #27211 from gottesmm/pr-9566688fe160ddaa5cdb741e119f513579ad8c4d

[ownership] Create a context structure for the linear lifetime checker.

jrose-apple committed:
[Frontend] Add a comment about 9e6d4db (#27169)

With wording help from David U.

hborla committed:
[Diagnostics] Add notes for the type_cannot_conform error that point
to the declaration that requires protocol conformance.

xedin committed:
[Diagnostics] Fix `missing argument` diagnostic to support closures with captures

xedin committed:
[Diagnostics] NFC: Remove "trailing" from method name for closures with missing args

`MissingArgumentsFailure::diagnoseClosure` can actually diagnose both
closures in argument positions as well as when their type comes from
context e.g. `let _: (Int) -> Void = {}`.

theblixguy committed:
[AST] Remove isTypeContext() check from hasValueSemantics() and move the responsibility to the caller

hborla committed:
[Diagnostics] Adjust the description of existential types for the
type_cannot_conform diagnostic message.

compnerd committed:
utils: bump to CMake 3.15.1

CMake 3.15.3 is the current stable release.  Update to 3.15.1 as CMake usually
has bug fixes in the first release.

hborla committed:
Merge remote-tracking branch 'upstream/master' into type-cannot-conform-diag

hborla committed:
[Test] Update Parse/confusables.swift with the "type cannot conform"
error message and note.

theblixguy committed:
[Typechecker] Remove a few unnecessary isTypeContext() checks as its implifed by isInstanceMember()

gottesmm committed:
[ownership] Add a new checkValue overload to LinearLifetimeChecker that validates the linear lifetime of the passed in value and returns bool to indicate success/failure.

This is the first in a series of patches that begin to hide the underlying
linear lifetime implementation underneath the facade of the linear lifetime
checker. I only updated the users that this applies to.

slavapestov committed:
Sema: Compute and store captures for default argument expressions

Commit:
[code-completion] Remove special handling for completion on the RHS of an assignment

This simplifies the code and prevents an assertion failure this code was
hitting computing the type relations between the result types and what it
determined as the expected type.

Resolves rdar://problem/53958454

xedin committed:
Merge pull request #27215 from xedin/remove-fn-dep-from-missing-args

[Diagnostics] A couple of clean-ups for `AddMissingArguments`

slavapestov committed:
SIL: Refactor type lowering computation of the effective generic signature

This also fixes a long-standing bug with default argument expressions
capturing generic parameters.

slavapestov committed:
Sema: Diagnose captures of dynamic 'Self' type from default argument expressions

We could actually allow this for local functions, but it's not
worth implementing that until the more general ability of local
function default arguments to capture values is implemented.

Fixes <rdar://problem/55119566>.

hborla committed:
[Diagnostics] In `MissingConformanceFailure`, if the non-conforming type
comes from an argument, use the argument expression as the anchor for
`diagnoseTypeCannotConform`.

Commit:
When ASTScope assertions fails, direct user to try disabling ASTScopes.

swift-ci committed:
Merge pull request #27219 from gottesmm/pr-76c8dc8ebfc577d7be7ce15c158337b575823c2a

Commit:
Add ASTScope_unreachable

Commit:
Add explanations to all asserts.

rintaro committed:
Revert "Merge pull request #27203 from CodaFi/movin-on-up"

This reverts commit 387d2a9aee35554d7ff62bba5a8082f3798f6d31, reversing
changes made to bf1ab6c29d2b895ac75a705d981e07f53ae15828.

atrick committed:
Add SIL SinkAddressProjections utility to avoid address phis.

Enable this utility during jump-threading in SimplifyCFG.

Ultimately, the SIL verifier should prevent all address-phis and we'll
need to use this utility in a few more places.

Fixes <rdar://problem/55320867> SIL verification failed: Unknown
formal access pattern: storage

theblixguy committed:
Merge pull request #27057 from theblixguy/unrevert/SR-11298

Unrevert "[Sema] Setter has incorrect mutating-ness inside class-constrained protocol extension"

jrose-apple committed:
[AST] Split FileUnit and its subclasses out of Module.h

Most of AST, Parse, and Sema deal with FileUnits regularly, but SIL
and IRGen certainly don't. Split FileUnit out into its own header to
cut down on recompilation times when something changes.

No functionality change.

jrose-apple committed:
[AST] Separate SourceFile from FileUnit.h

Like the last commit, SourceFile is used a lot by Parse and Sema, but
less so by the ClangImporter and (de)Serialization. Split it out to
cut down on recompilation times when something changes.

This commit does /not/ split the implementation of SourceFile out of
Module.cpp, which is where most of it lives. That might also be a
reasonable change, but the reason I was reluctant to is because a
number of SourceFile members correspond to the entry points in
ModuleDecl. Someone else can pick this up later if they decide it's a
good idea.

No functionality change.

theblixguy committed:
[Typechecker] Use an equal constraint between a property wrapper's value type and the pattern init

theblixguy committed:
[Test] Adds a test case

CodaFi committed:
Break some cycles

Use the interface type of the underlying type in the GSB in as many
cases as possible except for one important one: requirement signature
computation.  There, use the structural type so we don't wind up
recursively validating the requirements of an incomplete protocol.

ravikandhadai committed:
Merge pull request #27205 from ravikandhadai/constexpr-debug-stdlib

[Constant Evaluator] Support debug-stdlib, improve diagnostics and support `assertionFailure`

ravikandhadai committed:
Merge pull request #27208 from ravikandhadai/constexpr-fragment-check

[Constant Evaluator][Tests] Add tests to check constant evaluability of Swift code snippets

slavapestov committed:
SILGen: Fix crash when immediately-applied closure transitively captures generic parameters

slavapestov committed:
SIL: Compute lowered local captures of SILDeclRefs and not AnyFunctionRefs

Unfortuantely this commit is bigger than I would like but I couldn't think
of any reasonable ways to split it up.

The general idea here is that capture computation is now done for a
SILDeclRef and not an AnyFunctionRef. This allows SIL to represent the
captures of a default argument generator.

rintaro committed:
[SyntaxParses] Address memory issue in popTokenSyntax()

Don't return reference for local temporary object

rdar://problem/55421369

davezarzycki committed:
[SILOpt] NFC: fix -Wgnu-anonymous-struct warning

swift-ci committed:
Merge pull request #27231 from davezarzycki/pr27231

theblixguy committed:
Merge pull request #27226 from theblixguy/fix/SR-11381

[PropertyWrappers] Fix a bug with property type being overwritten with wrong type

jrose-apple committed:
[test] Run scale-tests on Linux too (#27223)

It's likely that these were listed as "REQUIRES: OS=macosx" to not
redundantly run them for iOS et al, but they don't take that long and
so it's more useful for Linux devs to be able to run them locally if
need be. Or to catch something that really is different on non-macOS.

compnerd committed:
Merge pull request #27214 from compnerd/bug-fixes

utils: bump to CMake 3.15.1

jrose-apple committed:
Merge pull request #27225 from jrose-apple/absolute-unit

[AST] Split FileUnit and SourceFile out of Module.h

varungandhi-apple committed:
[NFC] Replacing potentially confusing phrase "memory allocation promotion"

hborla committed:
Merge pull request #27176 from hborla/type-cannot-conform-diag

[Diagnostics] Generalize the "protocol type cannot conform" error to work for all types that cannot conform to protocols.

rintaro committed:
[SyntaxParse] Parse generic parameter clause and generic where clause

CodaFi committed:
Change a cycle condition

Not having the generic signature is the real culprit here.

slavapestov committed:
SILGen: Support for local function default argument captures

Fixes <https://bugs.swift.org/browse/SR-2189>, <rdar://problem/20796451>.

Commit:
Merge pull request #27218 from nathawes/r53958454-getFixedTypeRecursive-code-completion-crash

[code-completion] Remove special handling for completion on the RHS of an assignment

Commit:
Fix debug code

jckarter committed:
ReflectionContext: Remove platform #ifs for image format checks.

In principle, swift-reflection-* ought to work with cross-compiled binaries. Dispatch out to reading
MachO, PE, or ELF section metadata based on the magic of an image passed to `addImage` instead of
using #ifs to pick an implementation based on the host platform. (This still doesn't fully address
other host/target differences like word size or endianness, but is progress toward making the tool
target-agnostic.)

swift-ci committed:
Merge pull request #27234 from apple/vg-sil-docs

jckarter committed:
Merge pull request #27237 from jckarter/reflection-context-check-magic

ReflectionContext: Remove platform #ifs for image format checks.

slavapestov committed:
Merge pull request #27220 from slavapestov/default-arg-self-capture-diagnostic

Default arguments: implement captures for local functions, diagnose 'Self' for methods

CodaFi committed:
Merge pull request #27172 from CodaFi/aliasing-artifacts-and-noise-reduction-techniques

Kill validateDeclForNameLookup Harder

ravikandhadai committed:
[Constant Evaluator] Pass assert configuartion option to the
evaluator to precisely evaluate Builtin.assert_configuration.

Unify UnknownReason::Trap and UnknownReason::AssertionFailure error
values in the constant evaluator, now that we have 'condfail_message'
SIL instruction, which provides an error message for the traps.

rintaro committed:
[ASTGen] Adjust geneate() function signatures

'generate()' receives `const` reference to `SomeSyntax`. This should
prevent unnecessary reference counter operations.

rintaro committed:
[Parse] Simplify parseIdentifierDeclName()

As a preparation for syntax parsing.

beccadax committed:
[NFC] Note limitation of KnownProtocols

swift-ci committed:
Merge pull request #27228 from ravikandhadai/constexpr-oslog-constant-evaluable-check

slavapestov committed:
AST: Fix regression from SubstFlags::UseErrorTypes removal

Note that while the original crasher in the radar is gone, my reduced test
case triggers an IRGen crash on both 5.1 and master because of an unrelated
bug that appears to be related to protocol requirement signatures and
declaration ordering.

Fixes <rdar://problem/54952911>.

CodaFi committed:
[NFC] Address review feedback from #27172

- Clean up some errant formatting mistakes.
- Collapse some code that was duplicating computing the interface type.

swift-ci committed:
Merge pull request #27241 from CodaFi/well-i-should-have---run-clang-format

compnerd committed:
utils: pass `CMAKE_Swift_COMPILER`

This is the variable that the official CMake Swift support uses.  Define
this so that we can be forward compatible with CMake 3.15.

beccadax committed:
[NFC] [PrintAsObjC] Don’t cache NSCopyingType

ASTContext caches the decl and the decl caches the declared interface type, so there’s no performance win, and there isn’t a convenience win anymore either.

swift-ci committed:
Merge pull request #27244 from ravikandhadai/constexpr-oslog-constant-evaluable-check

beccadax committed:
[NFC] Convert ClangImporter use to getNSCopyingDecl()

rintaro committed:
Merge pull request #27230 from rintaro/syntaxparsse-rdar55421369

[SyntaxParse] Re-apply move-only ParsedRawSyntaxNode

Commit:
WIP lazy whole scopes

gottesmm committed:
[sil] Teach the SILVerifier how to validate that a builtin insts are always mapped to a known builtin.

The builtin instruction in SIL identifies the operation that it performs via a
simple string name. This commit changes the SILVerifier so that we validate that
any builtin inst can be mapped via its string name to a /true/ builtin value
decl.

Just tightening up the IR model.

atrick committed:
Merge pull request #27246 from atrick/fix-address-phi

Add SIL SinkAddressProjections utility to avoid address phis.

swift-ci committed:
Merge pull request #27251 from gottesmm/pr-92691592fd64691bf3f49ee560d0965c74a670c2

DougGregor committed:
[SILGen] Properly substitute contextual archetypes when generating assign_by_wrapper

Fixes SR-11303 / rdar://problem/54311335

gottesmm committed:
Merge pull request #27254 from gottesmm/pr-068c882956f6e9746cd2ea87e6aea7d6df8f1ac0

[builtin] Implement polymorphic builtins for all BUILTIN_BINARY_OPERA…

nate-chandler committed:
[cmpcodesize] Allowed args to specify dirs.

Previously, to specify the directories containing objects whose sizes
should be compared, environment variables had to be used.  With this
change, arguments (-o/--old-build-dir, -n/--new-build-dir) can be passed
to the cmpcodesize script directly.  If those arguments aren't
specified, the environment variables are sued as before.

slavapestov committed:
Add a couple of crashing test cases

slavapestov committed:
Merge pull request #27242 from slavapestov/subst-error-type-fix

AST: Fix regression from SubstFlags::UseErrorTypes removal

Commit:
Tag cycles w/ radar

gottesmm committed:
[polymorphic-builtins] Teach constant folding how to fold a polymorphic builtin to its static overload if the passed in type is a concrete builtin type

Specifically, this transforms:

builtin "generic_add"<Builtin.Vec4xInt32>(

  ->

builtin "add_Vec4xInt32"(

If we do not have a static overload for the type, we just leave the generic call
alone. If the generic builtin takes addresses as its arguments (i.e. 2x
in_guaranteed + 1x out), we load the arguments, evaluate the static overloaded
builtin and then store the result into the out parameter.

Commit:
Push range of extension scope after type, WIP

nate-chandler committed:
Merge pull request #27259 from nate-chandler/cmpcodesize-can-take-build-dirs-as-arguments

[cmpcodesize] Allowed args to specify dirs.

DougGregor committed:
Merge pull request #27257 from DougGregor/silgen-sr-11303

[SILGen] Properly substitute contextual archetypes when generating assign_by_wrapper

gottesmm committed:
[polymorphic-builtins] Teach dataflow diagnostics how to emit an error if it sees an unspecialized polymorphic builtin.

This will ensure that if an expert user is using this feature and makes a
mistake as a result of tweaking their code, they get an error. This will ensure
they do not ship and look into why this is happening.

This is not intended to be used by anyone except for expert stdlib users.

jrose-apple committed:
[docs] CToSwift: swift_name for all top-level declarations (#27227)

That is: types, globals, C functions, and enumerators, including a
discussion of import-as-member for each. Still more to come:
swift_name for properties, methods, and struct fields.

gottesmm committed:
Revert "[sil] Teach the SILVerifier how to validate that a builtin insts are always mapped to a known builtin."

This reverts commit e2a9a432209affcd5b427481795cc859bd5ae19d.

Going to put a cache in front of this to ensure we aren't bloating compile
time.

Conflicts:
	lib/SIL/SILVerifier.cpp

DougGregor committed:
Merge pull request #27262 from gottesmm/pr-f239c55c62358cc7f0c56e7588c784379003f933

Revert "[sil] Teach the SILVerifier how to validate that a builtin in…

jrose-apple committed:
[ClangImporter] Remove now-unnecessary special case for IUO importing

Back when IUO could be used anywhere in a type in Swift, we had a
special case for imported pointer types that their pointees were never
implicitly-unwrapped, even if they weren't nullability-audited on the
Objective-C side. Now that IUO can only be used in top-level positions
(SE-0054, only fully implemented last year), we don't need a special
case for this; all non-top-level optionals are never
implicitly-unwrapped.

No functionality change.

jrose-apple committed:
[ClangImporter] Update old comments and remove obsolete FIXMEs

jrose-apple committed:
[ClangImporter] Simplify typedef bridgeability checking

A typedef might get imported as an alias for a bridged type (String)
or for the original type (NSString), and a few parts of the importer
need to account for this. Simplify this logic based on how it's used
today.

jrose-apple committed:
[ClangImporter] Simplify a SmallVector fill to allocate up front

No functionality change.

jrose-apple committed:
[ClangImporter] Simplify helper method signature

No functionality change.

jrose-apple committed:
[ClangImporter] Simplify block pointer importing ever so slightly

No need to make a whole separate SwiftTypeConverter; just use the
one we already have.

gottesmm committed:
Merge pull request #27260 from gottesmm/pr-2b8fc4fe4ee6d645eeb2b8b243c07a8c2cd5d2da

[polymorphic-builtins] Teach constant folding how to fold a polymorph…

Commit:
Remove debug code

Commit:
De-experimentalify

Commit:
Silence a warning

xedin committed:
[ConstraintSystem] Introduce a fix for assignment type mismatch

Introduce a fix/diagnostic when there is a contextual mismatch
between source and destination types of the assignment e.g.:

```swift
var x: Int = 0
x = 4.0 // destination expects an `Int`, but source is a `Double`
```

Commit:
change compare-to-astScope-lookup to crosscheck-unqualified-lookup

compnerd committed:
Merge pull request #27245 from compnerd/peek-a-boo-i-see-you

utils: pass `CMAKE_Swift_COMPILER`

compnerd committed:
Update README.md

Split out the TensorFlow CI entries.  They were previously marked with a TensorFlow suffix, which obscured them in the rest of the community CI.  This makes them easier to find.

rintaro committed:
Merge pull request #27249 from rintaro/syntaxparse-genericparam

[SyntaxParse] Parse generic parameter clause and generic where clause

compnerd committed:
build: try harder to pass along CMAKE_Swift_COMPILER

This is needed for the CMake 3.15 based build.

slavapestov committed:
AST: Simplify SubstitutionMap::getProtocolSubstitutions()

gottesmm committed:
[polymorphic-builtin] Fix ASAN error.

slavapestov committed:
AST: ProtocolConformanceRef::subst() returns an invalid conformance instead of trapping

This is for consistency with Type::subst() which returns ErrorTypes.

slavapestov committed:
AST: Clean up ASTVerifier's archetype verification a bit

Commit:
Don't try to expand ASTSourceFileScope when eagerly building.

jckarter committed:
Reflection: Make RemoteRef template not depend on Runtime.

The only thing the Runtime affects is the width of the StoredPointer for the remote address, for
which storing a uint64_t ought to be enough for anyone we care about so far. This will make it
easier to store and use RemoteRefs in code that isn't or shouldn't ideally be templatized on
Runtime (such as TypeRefBuilder, and ultimately ReflectionContext, from the Reflection library.)

xedin committed:
[TypeChecker] Don't use contextual type for implicit `~=` type-check

Type-check of implicit `~=` has to be handled specially because
constraint system doesn't yet have a full access to pattern-matching
context where `~=` is used which would produce incorrect diagnostics.

slavapestov committed:
IDE: Pick off another usage of isConvertibleTo() with openArchetypes=true

Commit:
Ever expansion detection. WIP

Commit:
Fix typo

slavapestov committed:
AST: Remove LazyResolver::resolveProtocolEnvironment()

Commit:
Radar tagging

xedin committed:
[Diagnostics] Don't try to diagnose immutability failures in boolean context

CodaFi committed:
Make decl shadowing ask the right question

If it's going to validate the declaration, it wants to know if there's an interface type.  Computing the generic signature can be done on demand.

CodaFi committed:
Remove a dead cycle check

CodaFi committed:
Directly detect the cyclic case while computing conditional requirement

The general class of cycle here is when validation asks for the generic signature which triggers requirement checking which forces us to ask for the generic signature of the extension again.  We should look into a more principled solution.

See rdar://55263708

rintaro committed:
Revert "[SyntaxParse] Parse generic parameter clause and generic where clause"

CodaFi committed:
Remove resolveExtension and validateExtension

Push the relevant work this was doing down into the callers.  Most of them didn't want to validate the entire extension and force its generic parameters, they just wanted to validate the nominal type.

We must eagerly validate the generic signature of extensions, though.  This avoids a class of cycles where non-generic members of an extension with a generic signaure will call through to getGenericSignatureOfContext and force the generic signature anyways.  When this calls through to protocol witness matching, the signature can be recursively computed.

CodaFi committed:
Drop CheckingWithValidSignature from the validation state machine

Now that the generic signature is computable on demand, this predicate is doubly useless.  All of the callers intended to ask "hasInterfaceType" anyways.

jckarter committed:
Reflection: Move resolveRelativeField functionality onto RemoteRef

Resolving a direct relative reference given a RemoteRef doesn't need the MetadataReader,
since the offset should already be in the local buffer; we can add it to RemoteRef's
saved remote address and get a new remote address. Refactor the API to make as much as
possible of it available directly on RemoteRef.

rintaro committed:
Merge pull request #27278 from apple/revert-27249-syntaxparse-genericparam

Revert "[SyntaxParse] Parse generic parameter clause and generic where clause"

rintaro committed:
Revert "Revert "[SyntaxParse] Parse generic parameter clause and generic where clause""

This reverts commit 1584e87aa7dc92f1c4515da39367a63de1991e3a.

rintaro committed:
[Syntax] Fix Python lint failure

xedin committed:
Merge pull request #27107 from xedin/if-conditional-diag

[Diagnostics] Tailored diagnostic for "condition" expression

gottesmm committed:
Merge pull request #27272 from gottesmm/pr-30367b8cc5c6857ca99028d0972b5f07a394cbc0

[polymorphic-builtin] Fix ASAN error.

beccadax committed:
Merge pull request #27210 from brentdax/it-is-known

[NFC] Centralize lookup of NSCopying

jrose-apple committed:
[ClangImporter] Remove workaround for old NS_ERROR_ENUM import logic (#27267)

At one point the enum in an NS_ERROR_ENUM would get mapped to the
struct that the ClangImporter synthesized for it, rather than the code
enum that directly corresponded to what was declared in Objective-C.
That's not how the importer works anymore, though; we can trust that
the right type is picked without additional adjustments.

jrose-apple committed:
Merge pull request #27268 from jrose-apple/the-hit-new-game-sim-plify

[ClangImporter] Assorted NFC cleanups and simplifications

jrose-apple committed:
[AST] Shrink CaptureInfo down to a PointerIntPair (#27261)

If there are any actual captures, a CaptureInfoStorage will be
allocated in the ASTContext, matching the previous behavior of
allocating the array of captures there.

No functionality change. I /tried/ to change the functionality to
assert everywhere capture info was used but hadn't explicitly been
computed, but it turns out we do that /all over the place/ for any
declaration that's been synthesized, imported, or deserialized. I
left in some groundwork for someone to make this more explicit (or
requestify it) in the future.

compnerd committed:
stdlib: silence a unused function warning (NFC)

Pre-process away the unused function on Darwin.  NFC.

rintaro committed:
Merge pull request #27281 from rintaro/reapply-syntaxparse-genericparam

[SyntaxParse] Reapply Parse generic parameter clause and generic where clause

slavapestov committed:
Merge pull request #27280 from slavapestov/type-alias-sugar-prep

Preparations for subst() preserving sugared TypeAliasTypes

jrose-apple committed:
[ClangImporter] Remove now-unnecessary special case for IUO importing (#27265)

Back when IUO could be used anywhere in a type in Swift, we had a
special case for imported pointer types that their pointees were never
implicitly-unwrapped, even if they weren't nullability-audited on the
Objective-C side. Now that IUO can only be used in top-level positions
(SE-0054, only fully implemented last year), we don't need a special
case for this; all non-top-level optionals are never
implicitly-unwrapped.

No functionality change.

Commit:
[parser] Fix tuple type round-tripping issue

`public init(a: (b || c))` became `public init(a: )(b || c)` after
round-tripping.

Resolves rdar://problem/55280477

xedin committed:
Merge pull request #27283 from xedin/assign-conversion-failures

[ConstraintSystem] Introduce a fix for assignment type mismatch

CodaFi committed:
Merge pull request #26480 from CodaFi/makes-a-nominal-decl-healthy-wealthy-and-memberwise

[FB6918951] Expand "Make Memberwise Initializer" Refactoring Action

swift-ci committed:
Merge pull request #27285 from compnerd/tls-conditions

Commit:
Hacked not-so-lazy

Commit:
Rm temperature

CodaFi committed:
Merge pull request #27279 from CodaFi/eggs-tension-service

Kill validateExtension

davezarzycki committed:
[Testing] Workaroud glibc non-null attributes

davezarzycki committed:
Merge pull request #27288 from davezarzycki/pr27288

[Testing] Workaroud glibc non-null attributes

gottesmm committed:
Merge pull request #27284 from gottesmm/pr-492ee9b09850d10ed8d4b8bc54b10acc7c023ada

[polymorphic-builtins] Teach dataflow diagnostics how to emit an erro…

dan-zheng committed:
Add `-no-toolchain-stdlib-rpath` flag. (#27207)

Add `-no-toolchain-stdlib-rpath` flag: the negative version of
`-toolchain-stdlib-rpath`.

Make `-no-toolchain-stdlib-rpath` be the default: use `/usr/lib/swift` as
default RPATH on Darwin platforms instead of toolchain standard library.

Adapted from https://github.com/apple/swift/pull/27206.

tensorflow branch requires the opposite default (use toolchain standard
library as RPATH) because some stdlib modules like TensorFlow do not exist in
`/usr/lib/swift`.

compnerd committed:
Merge pull request #27271 from compnerd/swift-compiler

build: try harder to pass along CMAKE_Swift_COMPILER

CodaFi committed:
Merge pull request #27289 from CodaFi/ah-well-on-to-the-next-one

Remove resolveExtension and validateExtension

Commit:
Fix scope test

CodaFi committed:
[NFC] Remove unnecessary validation noise

Drop some callers to validateDecl and resolveDeclSignature that did not
actually need the interface type.

Commit:
Fully eager for printing, just eager enough for type-checking.

CodaFi committed:
Merge pull request #27295 from CodaFi/sign-on-the-dotted-line

[NFC] Remove unnecessary validation noise

Commit:
Ensure that after pushing start of ExtensionScope after extended type, the end is not before the beginning.

rintaro committed:
[Parse] DebuggerContextChange.h

Commit:
Add comment

Commit:
Comment out potential debugging code issue.

Commit:
Ensure that reexpansion does not lose scopes.

Commit:
renaming to ASTAncestor

Commit:
Use separate bool to record expansion

Commit:
Count AST ancestor scopes instead of children when last expanded

Commit:
Count AST ancestor scopes instead of children when last expanded

Commit:
Remove too-restrictive assertion

Commit:
cleanups

Commit:
Fix scope map test

Commit:
Turn on ASTScopes

Commit:
Merge pull request #27221 from davidungar/A-9-17-astscope-off

[NFC, NameLookup, ASTScope] Bug fix for eager scope tree construction & better failure messages

Commit:
Merge pull request #26573 from davidungar/A-7-9-astscope-no-nesting

[NameLookup ASTScope] On-by-default.

davezarzycki committed:
[AST] Fix debug typo

swift-ci committed:
Merge pull request #27297 from davezarzycki/pr27297

rintaro committed:
Merge pull request #27290 from rintaro/parse-identifierdeclname-simple

[Parse] Simplify parseIdentifierDeclName()

rintaro committed:
Merge pull request #27291 from rintaro/syntaxparses-astgen-signature

[ASTGen] Adjust geneate() function signatures

rintaro committed:
Merge pull request #27286 from nathawes/r55280477-tuple-type-syntax-round-trip

[parser] Fix invalid tuple type round-tripping issue

gottesmm committed:
[polymorphic-builtin] Rather than asserting in IRGen if we see a builtin that was not specialized, turn the builtin into a trap.

I forgot about this part of the design when I was working on this. To ensure
that the whole design works as expected, I included a small end-to-end test
using an experimental design for simd that uses polymorphic builtins that test
this functionally.

NOTE: The experimental design is only intended to exercise the code functionally.

rdar://48248417

shahmishal committed:
Merge pull request #27270 from apple/compnerd-tensorflow-status

Update README.md

gottesmm committed:
Fix no-asserts build.

xedin committed:
Merge pull request #26207 from owenv/new-vararg-conversion-diag

[Diagnostics] Improve diagnostic when attempting to pass an Array to a variadic argument

jckarter committed:
Merge pull request #27274 from jckarter/detemplate-remote-ref

Reflection: Make RemoteRef template not depend on Runtime.

gottesmm committed:
Merge pull request #27304 from gottesmm/pr-8fb189947015d45811a60087ae32efa918e96ca3

Fix no-asserts build.

jckarter committed:
Reflection: Remove vestigial Offset fields from ReflectionInfo

These are now always zero, because memory readers handle virtual address mapping.
The `swift_reflection_info_t` structure used by the C RemoteMirror API keeps
its offset fields because it's supposed to be a stable API, but we now assert that
the values are always zero.

Commit:
Update swift-stdlib-tool for Swift 5 (#27276)

swift-ci committed:
Merge pull request #27302 from gottesmm/pr-7a861faa03ec80b01ad9a212afe79e529fb6857e

gottesmm committed:
[sil] Add a new helper method ApplySite::getCalleeOperand().

I need this for some refactorings I am doing around BranchPropagatedUser.

We already had ApplySite::getCallee() which just returned the value of the
operand. I refactored it to call this instead.

DougGregor committed:
[Type checker] Validate subscript declaration before using it.

Fixes SR-1148 / rdar://problem/55303660.

gottesmm committed:
[sil] Add a new helper method CondBranchInst::isConditionOperand.

I am doing some refactoring on BranchPropagatedUser that is requiring me to work
a lot more with arguments and I found a need for this to avoid trafficing in
operand indices.

As part of this I also needed to write the helper getConditionOperand() to get
the underlying operand. I refactored the already preset getCondition() to use
this helper instead since it was just semantically that method + get operand's
value.

gottesmm committed:
[ownership] Delete dead code.

gottesmm committed:
[ownership] Only allow BranchPropagatedUser to be constructed from Operands.

This commit only changes how BranchPropagatedUser is constructed and does not
change the internal representation. This is a result of my noticing that
BranchPropagatedUser could also use an operand internally to represent its
state. To simplify how I am making the change, I am splitting the change into
two PRs that should be easy to validate:

1. A commit that maps how the various users of BranchPropagatedUser have been
constructing BPUs to a single routine that takes an Operand. This leaves
BranchPropagatedUser's internal state alone as well as its user the Linear
Lifetime Checker.

2. A second commit that changes the internal bits of the BranchPropagatedUser to
store an Operand instead of a PointerUnion.

This will allow me to use the first commit to validate the second.

swift-ci committed:
Merge pull request #27312 from gottesmm/pr-7a1cedad01d914933d33a3f8d7335f0abc06adc9

CodaFi committed:
Requestify getInterfaceType() in Name Only

Make getInterfaceType() call validateDecl on behalf of its clients.  In effect, this makes the interface type behave like a request.  It also means that its clients no longer need to perform a number of undesirable anti-patterns in order to sidestep the bizarre API contract validateDecl has at the moment

In particular, the following things are no longer necessary:

- Checking for a missing interface type then validating
- Validating the interface type then retrieving it
- Validating the interface type then retrieving a derived value

These anti-patterns will be removed in follow-up commits

CodaFi committed:
Port getInterfaceType() patterns in libAST

CodaFi committed:
Port getInterfaceType() patterns in libSema

CodaFi committed:
Port getInterfaceType() patterns in ancillary libraries

CodaFi committed:
Update SourceKit for getInterfaceType() changes

jrose-apple committed:
[ClangImporter] Remove dead code around UnsafePointer<Void> (#27310)

...which is how we imported pointers-to-void before UnsafeRawPointer
existed. No functionality change.

rintaro committed:
Merge pull request #27308 from rintaro/parse-debuggercontextchange

[Parse] Give own header to DebuggerContextChange. NFC

gottesmm committed:
[polymorphic-builtin] Disable experimental builtin test when optimization is enabled for now.

I am going to loop back around later and I want to unblock the package bots.

lorentey committed:
Merge pull request #27317 from gottesmm/pr-b5717067fec005e76089cd24f729fc2d1d2c4afb

[polymorphic-builtin] Disable experimental builtin test when optimiza…

jckarter committed:
Fix handling of null relative references in RemoteRef.

CodaFi committed:
Merge pull request #27314 from CodaFi/an-interface-to-the-interface

Requestify getInterfaceType() in Name Only

DougGregor committed:
Remove now-unnecessary validateDecl() call

jckarter committed:
Merge pull request #27318 from jckarter/remote-ref-null-relative

Fix handling of null relative references in RemoteRef.

compnerd committed:
Update README.md

The debian 9.5 builder has been down for 3 weeks.  Remove it until such time as it is able to build again.

shahmishal committed:
Merge pull request #27324 from apple/compnerd-community-ci-status

Update README.md

CodaFi committed:
[NFC] Drop some unnecesary interface type forcing

rintaro committed:
[SyntaxParse] Introduce CodeCompletionTypeSyntax

To represent a type with code completion.

  type? '.'? <code-completion-token>

This is "parser only" node which is not exposed to SwiftSyntax.
Using this, defer to set the parsed type to code-completion callbacks.

harlanhaskins committed:
[TBDGen] Set `installapi` flag only when building in installapi (#27315)

Previously, we would unconditionally set the `installapi` flag, which
hard errors when trying to merge with a TAPI-generated TBD file that
was generated during the install phase.

Fixes rdar://42406088

jckarter committed:
Merge pull request #27306 from jckarter/remove-reflection-offsets

Reflection: Remove vestigial Offset fields from ReflectionInfo

DougGregor committed:
Merge pull request #27320 from DougGregor/property-wrappers-subscript-multifile-sr11458

[Type checker] Validate subscript declaration before using it.

swift-ci committed:
Merge pull request #27326 from CodaFi/interface-to-face

rintaro committed:
Merge pull request #27325 from rintaro/syntaxparse-cctype

[SyntaxParse] Introduce CodeCompletionTypeSyntax

gottesmm committed:
Merge pull request #27316 from gottesmm/pr-634f227f60306a17bbad2b205f6bc0cea46541cf

[ownership] Only allow BranchPropagatedUser to be constructed from Operands.

dan-zheng committed:
Merge tag 'swift-DEVELOPMENT-SNAPSHOT-2019-09-24-a' into tensorflow-merge

Tag build swift-DEVELOPMENT-SNAPSHOT-2019-09-24-a

BradLarson committed:
Applying patch from TF-815 and updating the hashes for swift-apis and swift-quotes.

dan-zheng committed:
Fix test/Syntax/round_trip_parse_gen.swift.

Handle new `GenericRequirement` syntax node, added in:
https://github.com/apple/swift/pull/27249.

dan-zheng committed:
Workaround for test/SILOptimizer/constant_evaluable_subset_test.swift.

This test fails because `-enable-ownership-stripping-after-serialization`
was disabled for the standard library. Confirmed that re-enabling the flag
makes the test pass (revert f3c7cca91fd814fcbc93189a2f47edeca67c2ae4).

```
sil-opt: include/swift/SIL/SILBuilder.h:802:
swift::StoreInst *swift::SILBuilder::createStore(swift::SILLocation, swift::SILValue, swift::SILValue, swift::StoreOwnershipQualifier):
Assertion `(Qualifier != StoreOwnershipQualifier::Unqualified) || !hasOwnership() && "Unqualified inst in qualified function"' failed.
Stack dump:
1.      While running pass #259 SILModuleTransform "MandatoryInlining".
2.      While inlining SIL function "@$ss17FixedWidthIntegerPsE2lloiyxx_qd__tSzRd__lFZ".
 for .../test-linux-x86_64/SILOptimizer/Output/constant_evaluable_subset_test.swift.tmp/constant_evaluable_subset_test_silgen.sil:296:22
3.      While ...into SIL function "@$s37constant_evaluable_subset_test_silgen18leftShiftWithTraps1x1yS2i_SitF".
 for .../test-linux-x86_64/SILOptimizer/Output/constant_evaluable_subset_test.swift.tmp/constant_evaluable_subset_test_silgen.sil:282:54
```

Temporarily disabled test/SILOptimizer/constant_evaluable_subset_test.swift.
TF-799 tracks re-enabling the test.

dan-zheng committed:
Disable test/SILOptimizer/OSLogConstantEvaluableTest.swift.

This test fails because `-enable-ownership-stripping-after-serialization`
was disabled for the standard library.

TF-799 tracks re-enabling the test.

burmako committed:
Merge pull request #27453 from dan-zheng/disable-test-tf-799

Disable test/SILOptimizer/OSLogConstantEvaluableTest.swift.

dan-zheng committed:
Add ASTScope support for `@differentiable` attribute. (#27451)

`@differentiable` attribute where clauses may refer to generic
parameters from some generic context.

Without special ASTScope support for `@differentiable` attributes,
ASTScopeLookup.cpp logic tries to resolve the generic parameter
DeclNames in the where clause based on source location alone
(`ASTScopeImpl::findChildContaining`) and fails.

The fix is to add a special `DifferentiableAttributeScope`, mimicking
`SpecializeAttributeScope`. Every `@differentiable` attribute has its
own scope, derived from the declaration on which it is declared.
Unlike `@_specialize`, `@differentiable` may also be declared on
`AbstractStorageDecl` declarations (subscripts and variables).

Resolves TF-815.

`Decl::getSourceRangeIncludingAttrs` should not consider implicit
`@differentiable` attributes generated during `@differentiating`
attribute type-checking.

TF-835 tracks robust lowering for `@differentiating` attributes
that does not involve generating implicit `@differentiable` attributes,
circumventing this issue.

dan-zheng committed:
Disable test/AutoDiff/transposing_attr_type_checking.swift. (#27454)

TF-847 tracks re-enabling the test.
Use `XFAIL: *` to disable tests instead of commenting RUN lines.

burmako reviewed (approved):
LGTM after the merge conflict is fixed and tests pass.

dan-zheng committed:
Merge branch 'tensorflow' of github.com:apple/swift into tensorflow-merge

Updated utils/update_checkout/update-checkout-config.json.

Commit:
[tensorflow merge] fix qualified decl name parse (#27456)

dan-zheng committed:
Update checkout for lldb.

lldb requires an extra fix:
https://github.com/apple/swift-lldb/pull/2026/commits/8fc11a104c0930656469442bbc059eb1a2a52955

dan-zheng committed:
Merge pull request #27455 from dan-zheng/tensorflow-merge

Merge branch 'tensorflow' of github.com:apple/swift into tensorflow-merge

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please clean test tensorflow
